### ISSUE TITLE
Infrastructure for example generation in high-content modalities

### ISSUE THREAD

**Issue by cwbeitel** [projectclarify/pcml#97]
Various Q1 epics in area/machine-learning relate to video understanding. In previous quarter, infrastructure was built to support this and should be improved or replaced.

Key challenge in using very high performance accelerators to learn from video examples is that the input pipeline can easily become a bottleneck given the additional data that needs to be transferred and pre-processed. In principle the solution is simple - pre-compute as much pre-processing (e.g. data augmentation) as possible (so that it doesn’t need to be computed trainer-side) then store training-ready examples in the smallest size possible to minimize network bottlenecks.

Our current infrastructure for training video understanding models consists of three phases. The first is “extraction” which reads frame data from video files, down-samples it, and writes individual frames to a Google Cloud BigTable (CBT) table. A second phase, “continuous data generation”, samples frame sequences from the raw frame table, performs data augmentation (e.g. randomly augmenting color and contrast), and writes trainer-ready examples to a second CBT table. The third phase (occurring in parallel to the second once a minimum number of training examples have been generated) is the TPU training worker randomly sampling the second CBT table and using those examples to train the video understanding model (without doing any but the most minimal preprocessing trainer-side. How are these randomly sampled, you ask? TPU workers actually read the second CBT table in-order according to key but the example generation phase writes examples to this table randomly thereby avoiding the need for a costly shuffle step whether before or during training.

With such a design, example generation is at least decoupled from TPU training allowing us to avoid the challenge of input pipeline bottleneck for arbitrarily complex forms of data augmentation - provided we have enough example generation workers running to match the rate at which examples are consumed. Not doing so will result in overfitting.

Faster example generation means lower costs so in Q3 we built a prototype of a previous Python implementation in Golang, see here. This involves as much of the previous Python implementation remaining intact and calling the Golang prototype via a subprocess. Naturally, performance benefits were seen simply by using a more performant language with support for concurrency that is compatible with the currently available BigTable Client Library for Golang (this did not work well with the same for Python). Various additional enhancements included batching contiguous reads and writes from and to BigTable as well as sharding audio data so as to minimize the amount of data that needed to be read in each request (video was already sharded by virtue of being stored in individual frames).

This revised version of the example generation step allowed examples to be generated hundreds to thousands of times faster. But there are lots of things to improve with the prototype. One, at a systems level, is that there is currently nothing in place to throttle example generation based on the rate at which examples are being consumed. This involves doing a test run, looking at CBT logs to see the read rate in practice, then manually scaling the number of generation workers to approximately meet the consumption rate (including based on the generation rate over the number of workers used in the test run). Instead there could be a separate generation autoscaler that obtains CBT read and write metadata and autoscales generation workers to balance these. The value of such an implementation would not only allow costs on generation workers to be minimized but it would also allow the right rate of example generation throughout a hyperparameter tuning run - across which the rate of example consumption will vary as optimization speeds vary (for differently-sized and generally differently-configured models). Lastly, by creating abstractions to represent the process of generation and training at a higher level - integration testing of these two parts of the codebase would be improved.

Another area of improvement needed for the prototype is to generally improve the quality of the implementation (given the first version was written by a Python developer). While the prototype focused on example generation for images, video, and audio we will shortly wish to do the same in various other modalities and any modality-specific processing should be organized as such apart from a modality-independent core. But even within the current set of modalities there are various ways these can be constructed into examples as one can see by examining the various projects planned in SIG-ML for Q1 (video/video/video triplet correspondence, audio triplet correspondence, and both frame/audio and video/audio modality correspondence).

In addition to these there is a need for various forms of testing to confirm the correctness of the current implementation. While it was confirmed the current implementation achieves the desired example generation throughput we have not yet been able to train a high-quality model from examples generated in this way - plausibly because there is a bug somewhere in the implementation. One way to do such testing would be to simulate videos such as involving a ball bouncing around inside a square domain with an audio track that simply encodes the ball’s position. The trick is to construct a learning problem for which it is easy to verify if training examples have been generated correctly and having done so will be easy for a model to learn to solve.
